notebook multiselect revamp for actions

dhruvisompura opened the pull request:
Addresses #10116 and #10397

This PR revamps a good portion of our notebook cell actions to support multi-selections and ensure that undo/redo work as expected. 

Updated actions that now support multi-select properly with undo/redo:
- cut
- delete
- copy
- paste
- move cell up
- move cell down

I also fixed #10482 which fixes the "Run Below Cells" action to also include the current cell in its execution.

**Run Below Cells - BEFORE**

https://github.com/user-attachments/assets/ed45c886-46a7-470d-a416-9fbe38329f0e

**Run Below Cells - AFTER**

https://github.com/user-attachments/assets/2d1f91d6-855d-4974-9bac-f4929210f3bc

**Move Cells with Undo- AFTER**

https://github.com/user-attachments/assets/ec65dfff-77fd-4c87-b7e3-df0ba20fac41

**Cut/Copy/Paste/Delete with Undo- AFTER**

https://github.com/user-attachments/assets/e82354fc-2832-41bb-8852-4d12a5064737


### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes

@:positron-notebooks

<!--
  Positron team members: please add relevant e2e test tags, so the tests can be
  run when you open this pull request.

  - Instructions: https://github.com/posit-dev/positron/blob/main/test/e2e/README.md#pull-requests-and-test-tags
  - Available tags: https://github.com/posit-dev/positron/blob/main/test/e2e/infra/test-runner/test-tags.ts
-->


<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

dhruvisompura committed:
Simplify check that determines what cell action is run for

dhruvisompura committed:
Allow undo/redo to work for empty notebook

dhruvisompura committed:
Properly support multi-selected cells for delete (including undo/redo)

The delete cell action uses the `{ multiSelect: true }` option to handle multi-select scenarios. This causes `cell.delete()` to be called once for each selected cell, creating separate undo operations for each cell which is not correct for a multi-select scenario. When cells are deleted via multi-select, the deletes should happen as one atomic transaction and undo should bring back all deleted cells in the multi-select.

To fix these, we are using the `deleteCells` function instead of the `deleteCell` function which internally handled multiple cells and the undo/redo logic for all cells the operation should work.

I have also updated the action to extend NotebookAction2 to clearly communicate that this action relies on the notebook instance and is not run per cell.

dhruvisompura committed:
Update mulit-select supported  actions to use NotebookAction2 class

Actions that rely on a function which internally handled multi-select scenarios and undo/redo history should not use the CellAction2 class. The CellAction2 class for multi-select scenarios applies the action to each cell individually and tracks each operation in the undo/redo history as a separate transaction. This is confusing for users because the expectation is that all cells are part of a single transaction as far as undo/redo is concerned.

For example: When cutting 3 cells together, the undo operation should bring all 3 cells back. Prior to this change, undo would only bring the last cell back and the user would have to undo the operation per cell.

dhruvisompura committed:
Update paste cell action to operate off of active cell location

Update paste action to be a notebook level action since the internal pasteCells command handled multi-selection and utilizes the active cell for anchoring the pasted cells.

dhruvisompura committed:
Fix typo in e2e test name

dhruvisompura committed:
Clean up unused cell reference from move cell operations

dhruvisompura committed:
#10482 execute current cell in "Run Below Cells" command

dhruvisompura committed:
Support paste operation for empty notebooks

dhruvisompura committed:
Improve check that allows undo/redo to work for empty notebook

dhruvisompura commented on src/vs/workbench/contrib/positronNotebook/browser/positronNotebook.contribution.ts:
```diff
@@ -581,11 +575,9 @@ abstract class CellAction2 extends Action2 {
-			// Always check editing cell if actionBar is present (action bar items should work in edit mode).
-			// Otherwise, only check editing cell if editMode option is enabled.
-			const cell = getActiveCell(state) || ((isCellActionBarAction(this) || this.options?.editMode) ? getEditingCell(state) : undefined);
+			const cell = getActiveCell(state);
```
@seeM I updated the logic in here for determining the cell we should operate the cell action on when not in a mulit-selection scenario. I think having this just return the active cell makes sense since the active cell is the only one that can be in edit mode. 

Let me know if this seems right to you!

dhruvisompura commented on src/vs/workbench/contrib/positronNotebook/browser/positronNotebook.contribution.ts:
```diff
@@ -1088,20 +1080,23 @@ registerAction2(class extends CellAction2 {
+				when: ContextKeyExpr.or(
+					POSITRON_NOTEBOOK_EDITOR_CONTAINER_FOCUSED,
+					ContextKeyExpr.equals('activeEditor', POSITRON_NOTEBOOK_EDITOR_ID)
+				),
```
Pasting cells into a notebook needs to work regardless of the number of cells in the notebook. This means that all we really care about is the notebook being the active editor. I added that check. here. I don't think we need to check `POSITRON_NOTEBOOK_EDITOR_CONTAINER_FOCUSED` but I left it in. Do you think I should just remove this now?

dhruvisompura commented on src/vs/workbench/contrib/positronNotebook/browser/IPositronNotebookInstance.ts:
I made deleteCells public so we could use it for the delete action and renamed `moveCellUp`/`moveCellDown` to be plural since it supports multi-select.

dhruvisompura commented on test/e2e/tests/notebooks-positron/notebook-focus-and-selection.test.ts:
```diff
@@ -236,7 +236,7 @@ test.describe('Notebook Focus and Selection', {
 	});
 
-	test("`+ Code` and `+ Markdown` buttons insert the cell after the active cell and make it the new active cell)", async function ({ app }) {
+	test("`+ Code` and `+ Markdown` buttons insert the cell after the active cell and make it the new active cell", async function ({ app }) {
```
Just fixing a typo in the test name here

seeM reviewed (dismissed):
This works really well in my testing, very smooth! As you pointed out, we need to double-check with @nstrayer whether the changes affect any tools before merging

There are some considerations if we want our commands to be compatible with  VSCode counter parts but I don't think that's a priority right now and shouldn't be a heavy lift in future.

midleman committed:
Merge branch 'main' into feature/notebook-multiselect-cut-delete-undo-redo

midleman committed:
Merge branch 'main' into feature/notebook-multiselect-cut-delete-undo-redo

midleman committed:
unskip e2e tests

dhruvisompura committed:
Do not trigger paste action when a cell is in edit mode

dhruvisompura committed:
Simplify paste operation when clause

dhruvisompura committed:
Add support for empty notebook to context key

This fix updates the POSITRON_NOTEBOOK_EDITOR_CONTAINER_FOCUSED key to not be false when there are no cells in a notebook.

This ensures the paste action (and any other action that uses the POSITRON_NOTEBOOK_EDITOR_CONTAINER_FOCUSED context key) to only fire when the notebook editor has focus!

We can't rely on the POSITRON_NOTEBOOK_CELL_EDITOR_FOCUSED context key because it is false when there are no cells in the notebook and we can't rely on the activeEditor context keys because it is always true and causes the action to fire EVERY time a user presses v in an input.

dhruvisompura committed:
Hide notebook container focus outline

seeM reviewed (dismissed):
Code changes look good, good to merge once CI passes

dhruvisompura committed:
Update comment

dhruvisompura committed:
Run current cell as part of "Run Cells Above" cell action

dhruvisompura committed:
Remove CellAction2 class

seeM reviewed (approved):
Looks great, ty for the cleanup as well! E2e failures look unrelated to this change
